capi/include/hyper.h

```diff
@@ -0,0 +1,545 @@
+ task yields a `hyper_response *`.
+ */
+hyper_task *hyper_clientconn_send(hyper_clientconn *conn, hyper_request *req);
+
+/*
+ Free a `hyper_clientconn *`.
+ */
+void hyper_clientconn_free(hyper_clientconn *conn);
+
+/*
+ Creates a new set of HTTP clientconn options to be used in a handshake.
+ */
+hyper_clientconn_options *hyper_clientconn_options_new(void);
+
+/*
+ Free a `hyper_clientconn_options *`.
+ */
+void hyper_clientconn_options_free(hyper_clientconn_options *opts);
+
+/*
+ Set the client background task executor.
+
+ This does not consume the `options` or the `exec`.
+ */
+void hyper_clientconn_options_exec(hyper_clientconn_options *opts, const hyper_executor *exec);
+
+/*
+ Set the whether to use HTTP2.
+
+ Pass `0` to disable, `1` to enable.
+ */
+void hyper_clientconn_options_http2(hyper_clientconn_options *opts, int enabled);
+
+/*
+ Frees a `hyper_error`.
+ */
+void hyper_error_free(hyper_error *err);
+
+/*
+ Get an equivalent `hyper_code` from this error.
+ */
+hyper_code hyper_error_code(const hyper_error *err);
+
+/*
+ Print the details of this error to a buffer.
+
+ The `dst_len` value must be the maximum length that the buffer can
+ store.
+
+ The return value is number of bytes that were written to `dst`.
+ */
+size_t hyper_error_print(const hyper_error *err, uint8_t *dst, size_t dst_len);
+
+/*
+ Construct a new HTTP request.
+ */
+hyper_request *hyper_request_new(void);
+
+/*
+ Free an HTTP request if not going to send it on a client.
+ */
+void hyper_request_free(hyper_request *req);
+
+/*
+ Set the HTTP Method of the request.
+ */
+hyper_code hyper_request_set_method(hyper_request *req, const uint8_t *method, size_t method_len);
+
+/*
+ Set the URI of the request.
+ */
+hyper_code hyper_request_set_uri(hyper_request *req, const uint8_t *uri, size_t uri_len);
+
+/*
+ Gets a reference to the HTTP headers of this request
+
+ This is not an owned reference, so it should not be accessed after the
+ `hyper_request` has been consumed.
+ */
+hyper_headers *hyper_request_headers(hyper_request *req);
+
+/*
+ Set the body of the request.
+
+ The default is an empty body.
+
+ This takes ownership of the `hyper_body *`, you must not use it or
+ free it after setting it on the request.
+ */
+hyper_code hyper_request_set_body(hyper_request *req, hyper_body *body);
+
+/*
+ Free an HTTP response after using it.
+ */
+void hyper_response_free(hyper_response *resp);
+
+/*
+ Get the HTTP-Status code of this response.
+
+ It will always be within the range of 100-599.
+ */
+uint16_t hyper_response_status(const hyper_response *resp);
+
+/*
+ Get a pointer to the reason-phrase of this response.
+
+ This buffer is not null-terminated.
+
+ This buffer is owned by the response, and should not be used after
+ the response has been freed.
+
+ Use `hyper_response_reason_phrase_len()` to get the length of this
+ buffer.
+ */
+const uint8_t *hyper_response_reason_phrase(const hyper_response *resp);
+
```

Comment:
Are we any closer to this function being able to return the actually used reason instead of a fixed canonical version? This is after all information provided from the server that clients can use/inspect...

Reply:
I got distracted with other things, hoping that the canonical string was enough for most test cases. Grabbing the actual reason should be a quick add, aiming to get that in today.